src/api/callback.cc

```diff
@@ -158,20 +158,42 @@ MaybeLocal<Value> InternalMakeCallback(Environment* env,
     CHECK(!argv[i].IsEmpty());
 #endif
 
-  InternalCallbackScope scope(env, resource, asyncContext);
+  Local<Function> hook_cb = env->async_hooks_callback_trampoline();
+  int hookCount = 0;
+  int flags = 0;
+  if (!hook_cb.IsEmpty()) {
+    flags = InternalCallbackScope::kSkipAsyncHooks;
+    AsyncHooks* async_hooks = env->async_hooks();
+    hookCount = async_hooks->fields()[AsyncHooks::kBefore] +
+                async_hooks->fields()[AsyncHooks::kAfter];
+  }
+
+  InternalCallbackScope scope(env, resource, asyncContext, flags);
   if (scope.Failed()) {
     return MaybeLocal<Value>();
   }
 
   Local<Function> domain_cb = env->domain_callback();
   MaybeLocal<Value> ret;
-  if (asyncContext.async_id != 0 || domain_cb.IsEmpty()) {
-    ret = callback->Call(env->context(), recv, argc, argv);
-  } else {
+
+  if (asyncContext.async_id != 0 && hookCount != 0) {
+    std::vector<Local<Value>> args(3 + argc);
```

Comment:
Please use `MaybeStackBuffer` (e.g. similar to the transformation in c9d4949565c9a1e12cc1236cb4947ded88ebac3c), that way we can avoid the extra heap allocation in most cases

Reply:
Good call. I'll update that today. 👍